core/server/src/main/java/alluxio/worker/netty/DataServerWriteHandler.java

```diff
@@ -0,0 +1,338 @@
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.worker.netty;
+
+import alluxio.Configuration;
+import alluxio.Constants;
+import alluxio.PropertyKey;
+import alluxio.network.protocol.RPCMessage;
+import alluxio.network.protocol.RPCProtoMessage;
+import alluxio.network.protocol.databuffer.DataBuffer;
+import alluxio.proto.dataserver.Protocol;
+
+import com.google.common.base.Preconditions;
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.util.ReferenceCountUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.LinkedList;
+import java.util.Queue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.locks.ReentrantLock;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * This class handles {@link alluxio.proto.dataserver.Protocol.WriteRequest}s.
+ *
+ * Protocol: Check {@link alluxio.client.block.stream.NettyPacketWriter} for more information.
+ * 1. The netty channel handler streams packets from the channel and buffers them. The netty
+ *    reader is paused if the buffer is full by turning off the auto read, and is resumed when
+ *    the buffer is not full.
+ * 2. The {@link PacketWriter} polls packets from the buffer and writes to the block worker. The
+ *    writer becomes inactive if there is nothing on the buffer to free up the executor. It is
+ *    resumed when the buffer becomes non-empty.
+ * 3. When an error occurs, the channel is closed. All the buffered packets are released when the
+ *    channel is deregistered.
+ */
+@NotThreadSafe
+public abstract class DataServerWriteHandler extends ChannelInboundHandlerAdapter {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  private static final int MAX_PACKETS_IN_FLIGHT =
+      Configuration.getInt(PropertyKey.WORKER_NETWORK_NETTY_WRITER_BUFFER_SIZE_PACKETS);
+
+  /** The executor service to run the {@link PacketWriter}s. */
+  private final ExecutorService mPacketWriterExecutor;
+
+  private ReentrantLock mLock = new ReentrantLock();
+  /** The buffer for packets read from the channel. */
+  @GuardedBy("mLock")
+  private Queue<ByteBuf> mPackets = new LinkedList<>();
+  /** Set to true if the packet writer is active. */
+  @GuardedBy("mLock")
+  private boolean mPacketWriterActive = false;
+
+  protected volatile WriteRequestInternal mRequest = null;
+
+  protected abstract class WriteRequestInternal implements Closeable {
```

Comment:
I prefer not as I need to reference WriteRequestInternal above. It sounds strange to have an instance of AbstractXX